text/src/autogluon/text/automm/data/datamodule.py

```diff
@@ -0,0 +1,103 @@
+from .dataset import BaseDataset
+from .collator import Dict
+from ..constants import TRAIN, VAL, TEST, PREDICT
+from .preprocess_dataframe import MultiModalFeaturePreprocessor
+
+
+class BaseDataModule(LightningDataModule):
+    def __init__(
+            self,
+            df_preprocessor: MultiModalFeaturePreprocessor,
+            data_processors: dict,
+            per_gpu_batch_size: int,
+            num_workers: int,
+            train_data: Optional[pd.DataFrame] = None,
+            val_data: Optional[pd.DataFrame] = None,
+            test_data: Optional[pd.DataFrame] = None,
+            predict_data: Optional[pd.DataFrame] = None,
+    ):
+        super().__init__()
+        self.prepare_data_per_node = True
+
+        self.df_preprocessor = df_preprocessor
+        self.data_processors = data_processors
+        self.per_gpu_batch_size = per_gpu_batch_size
+        self.num_workers = num_workers
+        self.train_data = train_data
+        self.val_data = val_data
+        self.test_data = test_data
+        self.predict_data = predict_data
+
+    def set_dataset(self, split):
+        data_split = getattr(self, f"{split}_data")
+        dataset = BaseDataset(
+            data=data_split,
+            preprocessor=self.df_preprocessor,
+            processors=self.data_processors,
+            is_training=split == TRAIN,
+        )
+
+        setattr(self, f"{split}_dataset", dataset)
+
+    def setup(self, stage):
+        if stage == "fit":
+            self.set_dataset(TRAIN)
+            self.set_dataset(VAL)
+        elif stage == "test":
+            self.set_dataset(TEST)
+        elif stage == "predict":
+            self.set_dataset(PREDICT)
+        else:
+            raise ValueError(f"Unknown stage {stage}")
+
+    def train_dataloader(self):
+        loader = DataLoader(
+            self.train_dataset,
+            batch_size=self.per_gpu_batch_size,
+            num_workers=self.num_workers,
+            pin_memory=False,
+            collate_fn=self.get_collate_fn(),
+        )
+        return loader
+
+    def val_dataloader(self):
+        loader = DataLoader(
+            self.val_dataset,
+            batch_size=self.per_gpu_batch_size,
+            num_workers=self.num_workers,
+            pin_memory=False,
+            collate_fn=self.get_collate_fn(),
+        )
+        return loader
+
+    def test_dataloader(self):
+        loader = DataLoader(
+            self.test_dataset,
+            batch_size=self.per_gpu_batch_size,
+            num_workers=self.num_workers,
+            pin_memory=False,
+            collate_fn=self.get_collate_fn(),
+        )
+        return loader
+
+    def predict_dataloader(self):
+        loader = DataLoader(
+            self.predict_dataset,
+            batch_size=self.per_gpu_batch_size,
+            num_workers=self.num_workers,
+            pin_memory=False,
+            collate_fn=self.get_collate_fn(),
+        )
+        return loader
```

Comment:
Code duplication for the `DataLoader` construction with only 1st parameter changing - consider extracting a function

Reply:
The {train/val/test/predict}_dataloader is registered by Pytorch Lightning. Since we don't know when extactly these methods will be called, we should use it in this way.
Refer to https://pytorch-lightning.readthedocs.io/en/latest/extensions/datamodules.html#lightningdatamodule-api